Build process doesn't end

# Bug report

## Describe the bug

I keep trying to run "next build", and it won't work. I've tried deleting the .next folder, and I've ran both yarn and npm (after deleting yarn.lock, and package-lock, respectively)
## To Reproduce

Checkout 

1.Clone  https://github.com/AmherstStudent/amherst-student-site/tree/code-clean
2. Run yarn && yarn build
3. See error

## Expected behavior

Finished build process


## System information

- Windows
- Version of Next.js:  v8. canary

```

./components/Article/Image.js
Module not found: Can't resolve '../Basics/Card' in '/home/jozef-piecyk-10clouds/nextjs/amherst-student-site/client/components/Article'

```

You're trying to import file `Card` (Uppercase) while you only have `card.js` (lowercase) in the directory you're importing module from. After fixing this I got error that it cannot resolve components/Article/AuthorCard, again filename is authorcard. Make sure that filename and path in import is the same. You can use eslint or some extension to help you with this (like PathIntellisense for VSCode)

--- a/packages/next/build/index.ts
+++ b/packages/next/build/index.ts
@@ -15,7 +15,7 @@ import loadConfig, {
 import nanoid from 'next/dist/compiled/nanoid/index.js'
 import path from 'path'
 import { promisify } from 'util'
-import workerFarm from 'worker-farm'
+import Worker from 'jest-worker'
 
 import formatWebpackMessages from '../client/dev/error-overlay/format-webpack-messages'
 import { recursiveDelete } from '../lib/recursive-delete'
@@ -208,13 +208,9 @@ export default async function build(dir: string, conf = null): Promise<void> {
   const staticCheckSema = new Sema(config.experimental.cpus, {
     capacity: pageKeys.length,
   })
-  const staticCheckWorkers = workerFarm(
-    {
-      maxConcurrentWorkers: config.experimental.cpus,
-    },
-    staticCheckWorker,
-    ['default']
-  )
+  const staticCheckWorkers = new Worker(staticCheckWorker, {
+    numWorkers: config.experimental.cpus,
+  })
 
   await Promise.all(
     pageKeys.map(async page => {
@@ -270,14 +266,9 @@ export default async function build(dir: string, conf = null): Promise<void> {
       if (nonReservedPage) {
         try {
           await staticCheckSema.acquire()
-          const result: any = await new Promise((resolve, reject) => {
-            staticCheckWorkers.default(
-              { serverBundle, runtimeEnvConfig },
-              (error: Error | null, result: any) => {
-                if (error) return reject(error)
-                resolve(result || {})
-              }
-            )
+          let result: any = await (staticCheckWorkers as any).default({
+            serverBundle,
+            runtimeEnvConfig,
           })
           staticCheckSema.release()
 
@@ -301,8 +292,7 @@ export default async function build(dir: string, conf = null): Promise<void> {
       pageInfos.set(page, { size, chunks, serverBundle, static: isStatic })
     })
   )
-
-  workerFarm.end(staticCheckWorkers)
+  staticCheckWorkers.end()
 
   if (invalidPages.size > 0) {
     throw new Error(

--- a/packages/next/build/index.ts
+++ b/packages/next/build/index.ts
@@ -210,6 +210,7 @@ export default async function build(dir: string, conf = null): Promise<void> {
   })
   const staticCheckWorkers = new Worker(staticCheckWorker, {
     numWorkers: config.experimental.cpus,
+    enableWorkerThreads: true,
   })
 
   await Promise.all(

--- a/packages/next/build/index.ts
+++ b/packages/next/build/index.ts
@@ -1,4 +1,3 @@
-import { Sema } from 'async-sema'
 import chalk from 'chalk'
 import fs from 'fs'
 import mkdirpOrig from 'mkdirp'
@@ -204,9 +203,6 @@ export default async function build(dir: string, conf = null): Promise<void> {
 
   process.env.NEXT_PHASE = PHASE_PRODUCTION_BUILD
 
-  const staticCheckSema = new Sema(config.experimental.cpus, {
-    capacity: pageKeys.length,
-  })
   const staticCheckWorkers = new Worker(staticCheckWorker, {
     numWorkers: config.experimental.cpus,
   })
@@ -264,12 +260,10 @@ export default async function build(dir: string, conf = null): Promise<void> {
 
       if (nonReservedPage) {
         try {
-          await staticCheckSema.acquire()
           let result: any = await (staticCheckWorkers as any).default({
             serverBundle,
             runtimeEnvConfig,
           })
-          staticCheckSema.release()
 
           if (result.isHybridAmp) {
             hybridAmpPages.add(page)
@@ -284,7 +278,6 @@ export default async function build(dir: string, conf = null): Promise<void> {
         } catch (err) {
           if (err.message !== 'INVALID_DEFAULT_EXPORT') throw err
           invalidPages.add(page)
-          staticCheckSema.release()
         }
       }
 

--- a/packages/next/build/static-checker.ts
+++ b/packages/next/build/static-checker.ts
@@ -1,14 +1,9 @@
 import { isPageStatic } from './utils'
 
-export default function worker(
-  options: any,
-  callback: (err: Error | null, data?: any) => void
-) {
-  try {
-    const { serverBundle, runtimeEnvConfig } = options || ({} as any)
-    const result = isPageStatic(serverBundle, runtimeEnvConfig)
-    callback(null, result)
-  } catch (error) {
-    callback(error)
-  }
+export default function worker(options: {
+  serverBundle: string
+  runtimeEnvConfig: any
+}) {
+  const { serverBundle, runtimeEnvConfig } = options || ({} as any)
+  return isPageStatic(serverBundle, runtimeEnvConfig)
 }

--- a/packages/next/build/webpack/plugins/terser-webpack-plugin/src/TaskRunner.js
+++ b/packages/next/build/webpack/plugins/terser-webpack-plugin/src/TaskRunner.js
@@ -29,14 +29,8 @@ export default class TaskRunner {
     }
 
     if (this.maxConcurrentWorkers > 1) {
-      let enableWorkerThreads = false
-      try {
-        require('worker_threads')
-        enableWorkerThreads = true
-      } catch (_) {}
-
       this.workers = new Worker(worker, {
-        enableWorkerThreads,
+        enableWorkerThreads: true,
         numWorkers: this.maxConcurrentWorkers,
       })
       this.boundWorkers = options => this.workers.default(options)

--- a/packages/next/build/webpack/plugins/terser-webpack-plugin/src/TaskRunner.js
+++ b/packages/next/build/webpack/plugins/terser-webpack-plugin/src/TaskRunner.js
@@ -1,9 +1,8 @@
 import { join } from 'path'
 import minify from './minify'
 import { promisify } from 'util'
-import workerFarm from 'worker-farm'
+import Worker from 'jest-worker'
 import { writeFile, readFile } from 'fs'
-import serialize from 'serialize-javascript'
 import { Sema } from 'async-sema'
 import mkdirp from 'mkdirp'
 
@@ -16,7 +15,6 @@ export default class TaskRunner {
     if (cache) {
       mkdirp.sync((this.cacheDir = join(distDir, 'cache', 'next-minifier')))
     }
-
     // In some cases cpus() returns undefined
     // https://github.com/nodejs/node/issues/19022
     this.maxConcurrentWorkers = cpus
@@ -31,30 +29,19 @@ export default class TaskRunner {
     }
 
     if (this.maxConcurrentWorkers > 1) {
-      const workerOptions =
-        process.platform === 'win32'
-          ? {
-              maxConcurrentWorkers: this.maxConcurrentWorkers,
-              maxConcurrentCallsPerWorker: 1,
-            }
-          : { maxConcurrentWorkers: this.maxConcurrentWorkers }
-      this.workers = workerFarm(workerOptions, worker)
-      this.boundWorkers = (options, cb) => {
-        try {
-          this.workers(serialize(options), cb)
-        } catch (error) {
-          // worker-farm can fail with ENOMEM or something else
-          cb(error)
-        }
-      }
+      let enableWorkerThreads = false
+      try {
+        require('worker_threads')
+        enableWorkerThreads = true
+      } catch (_) {}
+
+      this.workers = new Worker(worker, {
+        enableWorkerThreads,
+        numWorkers: this.maxConcurrentWorkers,
+      })
+      this.boundWorkers = options => this.workers.default(options)
     } else {
-      this.boundWorkers = (options, cb) => {
-        try {
-          cb(null, minify(options))
-        } catch (error) {
-          cb(error)
-        }
-      }
+      this.boundWorkers = async options => minify(options)
     }
 
     let toRun = tasks.length
@@ -71,20 +58,19 @@ export default class TaskRunner {
 
     tasks.forEach((task, index) => {
       const cachePath = this.cacheDir && join(this.cacheDir, task.cacheKey)
-
-      const enqueue = () => {
-        this.boundWorkers(task, (error, data) => {
-          const result = error ? { error } : data
+      const enqueue = async () => {
+        try {
+          const result = await this.boundWorkers(task)
           const done = () => step(index, result)
-
-          if (this.cacheDir && !result.error) {
-            writeFileP(cachePath, JSON.stringify(data), 'utf8')
+          if (cachePath) {
+            writeFileP(cachePath, JSON.stringify(result), 'utf8')
               .then(done)
               .catch(done)
-          } else {
-            done()
           }
-        })
+        } catch (error) {
+          console.error('Error occurred minifying', error)
+          step(index, error)
+        }
       }
 
       this.sema.acquire().then(() => {
@@ -101,7 +87,7 @@ export default class TaskRunner {
 
   exit() {
     if (this.workers) {
-      workerFarm.end(this.workers)
+      this.workers.end()
     }
   }
 }

--- a/packages/next/build/webpack/plugins/terser-webpack-plugin/src/TaskRunner.js
+++ b/packages/next/build/webpack/plugins/terser-webpack-plugin/src/TaskRunner.js
@@ -3,10 +3,9 @@ import minify from './minify'
 import { promisify } from 'util'
 import Worker from 'jest-worker'
 import { writeFile, readFile } from 'fs'
-import { Sema } from 'async-sema'
 import mkdirp from 'mkdirp'
 
-const worker = require.resolve('./worker')
+const worker = require.resolve('./minify')
 const writeFileP = promisify(writeFile)
 const readFileP = promisify(readFile)
 
@@ -18,7 +17,6 @@ export default class TaskRunner {
     // In some cases cpus() returns undefined
     // https://github.com/nodejs/node/issues/19022
     this.maxConcurrentWorkers = cpus
-    this.sema = new Sema(cpus * 3)
   }
 
   run(tasks, callback) {
@@ -47,7 +45,6 @@ export default class TaskRunner {
     let toRun = tasks.length
     const results = []
     const step = (index, data) => {
-      this.sema.release()
       toRun -= 1
       results[index] = data
 
@@ -68,20 +65,17 @@ export default class TaskRunner {
               .catch(done)
           }
         } catch (error) {
-          console.error('Error occurred minifying', error)
-          step(index, error)
+          step(index, { error })
         }
       }
 
-      this.sema.acquire().then(() => {
-        if (this.cacheDir) {
-          readFileP(cachePath, 'utf8')
-            .then(data => step(index, JSON.parse(data)))
-            .catch(() => enqueue())
-        } else {
-          enqueue()
-        }
-      })
+      if (this.cacheDir) {
+        readFileP(cachePath, 'utf8')
+          .then(data => step(index, JSON.parse(data)))
+          .catch(() => enqueue())
+      } else {
+        enqueue()
+      }
     })
   }
 

--- a/packages/next/build/webpack/plugins/terser-webpack-plugin/src/worker.js
+++ b/packages/next/build/webpack/plugins/terser-webpack-plugin/src/worker.js
@@ -1,21 +1,2 @@
 import minify from './minify'
-
-module.exports = (options, callback) => {
-  try {
-    // 'use strict' => this === undefined (Clean Scope)
-    // Safer for possible security issues, albeit not critical at all here
-    // eslint-disable-next-line no-new-func, no-param-reassign
-    options = new Function(
-      'exports',
-      'require',
-      'module',
-      '__filename',
-      '__dirname',
-      `'use strict'\nreturn ${options}`
-    )(exports, require, module, __filename, __dirname)
-
-    callback(null, minify(options))
-  } catch (errors) {
-    callback(errors)
-  }
-}
+export default options => minify(options)

--- a/packages/next/build/webpack/plugins/terser-webpack-plugin/src/worker.js
+++ b/packages/next/build/webpack/plugins/terser-webpack-plugin/src/worker.js
@@ -1,2 +1,0 @@
-import minify from './minify'
-export default options => minify(options)

--- a/packages/next/package.json
+++ b/packages/next/package.json
@@ -80,6 +80,7 @@
     "find-up": "4.0.0",
     "fork-ts-checker-webpack-plugin": "1.3.4",
     "fresh": "0.5.2",
+    "jest-worker": "24.9.0",
     "launch-editor": "2.2.1",
     "loader-utils": "1.2.3",
     "mkdirp": "0.5.1",
@@ -88,7 +89,6 @@
     "prop-types-exact": "1.2.0",
     "react-error-overlay": "5.1.6",
     "react-is": "16.8.6",
-    "serialize-javascript": "1.7.0",
     "source-map": "0.6.1",
     "string-hash": "1.1.3",
     "strip-ansi": "5.2.0",
@@ -101,8 +101,7 @@
     "webpack": "4.39.0",
     "webpack-dev-middleware": "3.7.0",
     "webpack-hot-middleware": "2.25.0",
-    "webpack-sources": "1.3.0",
-    "worker-farm": "1.7.0"
+    "webpack-sources": "1.3.0"
   },
   "peerDependencies": {
     "react": "^16.6.0",

